File: platform/services/ml/app/storage.py
"""MinIO/S3 helpers. Keys are 'bucket/path' strings platform-wide."""
import boto3

from .config import settings

_s3 = None


def s3():
    global _s3
    if _s3 is None:
        _s3 = boto3.client(
            "s3",
            endpoint_url=settings.s3_endpoint,
            aws_access_key_id=settings.s3_access_key,
            aws_secret_access_key=settings.s3_secret_key,
        )
    return _s3


def split_key(full_key: str) -> tuple[str, str]:
    bucket, _, key = full_key.partition("/")
    return bucket, key


def download(full_key: str, dest_path: str) -> None:
    bucket, key = split_key(full_key)
    s3().download_file(bucket, key, dest_path)

File: platform/services/render/src/storage.ts
import { createReadStream, createWriteStream } from "node:fs";
import { pipeline } from "node:stream/promises";
import { Readable } from "node:stream";
import { GetObjectCommand, PutObjectCommand, S3Client } from "@aws-sdk/client-s3";

export const BUCKETS = {
  sources: process.env.S3_BUCKET_SOURCES ?? "sources",
  artifacts: process.env.S3_BUCKET_ARTIFACTS ?? "artifacts",
  renders: process.env.S3_BUCKET_RENDERS ?? "renders",
};

export const s3 = new S3Client({
  endpoint: process.env.S3_ENDPOINT ?? "http://localhost:9000",
  region: process.env.S3_REGION ?? "us-east-1",
  forcePathStyle: true,
  credentials: {
    accessKeyId: process.env.S3_ACCESS_KEY ?? "onepct",
    secretAccessKey: process.env.S3_SECRET_KEY ?? "onepct-secret",
  },
});

/** Keys are "bucket/key/path" strings throughout the platform. */
export function splitKey(fullKey: string): { bucket: string; key: string } {
  const idx = fullKey.indexOf("/");
  return { bucket: fullKey.slice(0, idx), key: fullKey.slice(idx + 1) };
}

export async function downloadToFile(fullKey: string, destPath: string): Promise<void> {
  const { bucket, key } = splitKey(fullKey);
  const res = await s3.send(new GetObjectCommand({ Bucket: bucket, Key: key }));
  await pipeline(res.Body as Readable, createWriteStream(destPath));
}

export async function uploadFile(fullKey: string, srcPath: string, contentType?: string): Promise<void> {
  const { bucket, key } = splitKey(fullKey);
  await s3.send(
    new PutObjectCommand({ Bucket: bucket, Key: key, Body: createReadStream(srcPath), ContentType: contentType })
  );
}

export async function uploadBuffer(fullKey: string, body: Buffer | string, contentType?: string): Promise<void> {
  const { bucket, key } = splitKey(fullKey);
  await s3.send(new PutObjectCommand({ Bucket: bucket, Key: key, Body: body, ContentType: contentType }));
}
